File: agent-runner/src/mcp/tools/pr.ts
// agent-runner/src/mcp/tools/pr.ts
import { tool } from "@anthropic-ai/claude-agent-sdk";
import { z } from "zod";
import type { WorkspaceContext } from "../context.js";
import { fetchWithRetry, formatFetchError, buildHeaders, BACKEND_URL } from "./fetch-utils.js";

// Matches GitHub PR URLs: https://github.com/owner/repo/pull/123
const PR_URL_PATTERN = /^https:\/\/github\.com\/[^/]+\/[^/]+\/pull\/\d+$/;

export function createPRTools(context: WorkspaceContext) {
  return [
    tool(
      "report_pr_created",
      "Report that a pull request was created for this session. Call this AFTER successfully creating a PR with `gh pr create` or any other method. This ensures the PR is immediately tracked in the ChatML UI.",
      {
        prNumber: z.coerce.number().int().positive().describe("The PR number (e.g., 123)"),
        prUrl: z.string().describe("The full PR URL (e.g., https://github.com/owner/repo/pull/123)"),
      },
      async ({ prNumber, prUrl }) => {
        if (!PR_URL_PATTERN.test(prUrl)) {
          return {
            content: [{
              type: "text" as const,
              text: `Invalid PR URL: expected a GitHub PR URL like https://github.com/owner/repo/pull/123, got: ${prUrl}`,
            }],
          };
        }

        try {
          const response = await fetchWithRetry(
            `${BACKEND_URL}/api/repos/${context.workspaceId}/sessions/${context.sessionId}/pr/report`,
            {
              method: "POST",
              headers: buildHeaders(true),
              body: JSON.stringify({ prNumber, prUrl }),
            }
          );

          if (!response.ok) {
            const error = await response.text();
            return {
              content: [{
                type: "text" as const,
                text: `Failed to report PR: ${response.status} ${error}`,
              }],
            };
          }

          return {
            content: [{
              type: "text" as const,
              text: `Reported PR #${prNumber} to ChatML. The PR badge will appear in the sidebar.`,
            }],
          };
        } catch (error) {
          return {
            content: [{
              type: "text" as const,
              text: `Error reporting PR: ${formatFetchError(error)}. The PR may not appear in the ChatML sidebar immediately, but the PR watcher will pick it up.`,
            }],
          };
        }
      },
      { annotations: { readOnlyHint: false } }
    ),

    tool(
      "report_pr_merged",
      "Report that a pull request was merged for this session. Call this AFTER successfully merging a PR with `gh pr merge` or any other method. This updates the session status in ChatML.",
      {
        prNumber: z.coerce.number().int().positive().optional().describe("The PR number that was merged (optional if the session already has a PR associated)"),
      },
      async ({ prNumber }) => {
        try {
          const response = await fetchWithRetry(
            `${BACKEND_URL}/api/repos/${context.workspaceId}/sessions/${context.sessionId}/pr/report-merge`,
            {
              method: "POST",
              headers: buildHeaders(true),
              body: JSON.stringify({ prNumber }),
            }
          );

          if (!response.ok) {
            const error = await response.text();
            return {
              content: [{
                type: "text" as const,
                text: `Failed to report PR merge: ${response.status} ${error}`,
              }],
            };
          }

          return {
            content: [{
              type: "text" as const,
              text: prNumber
                ? `Reported PR #${prNumber} merge to ChatML. Session status will update shortly.`
                : `Reported PR merge to ChatML. Session status will update shortly.`,
            }],
          };
        } catch (error) {
          // Best-effort notification — the PR watcher will detect merges via polling.
          return {
            content: [{
              type: "text" as const,
              text: prNumber
                ? `Failed to report PR #${prNumber} merge to backend (${formatFetchError(error)}), but the PR watcher will detect the merge automatically.`
                : `Failed to report PR merge to backend (${formatFetchError(error)}), but the PR watcher will detect the merge automatically.`,
            }],
          };
        }
      },
      { annotations: { readOnlyHint: false } }
    ),
  ];
}

File: backend/loop/chatml/pr.go
package chatml

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"

	"github.com/chatml/chatml-core/tool"
)

var prURLPattern = regexp.MustCompile(`^https://github\.com/[^/]+/[^/]+/pull/\d+$`)

// --- report_pr_created ---

type reportPRCreatedTool struct {
	svc *Services
	ctx *ToolContext
}

func (t *reportPRCreatedTool) Name() string           { return "mcp__chatml__report_pr_created" }
func (t *reportPRCreatedTool) IsConcurrentSafe() bool { return true }
func (t *reportPRCreatedTool) DeferLoading() bool     { return true }
func (t *reportPRCreatedTool) Description() string {
	return "Report that a pull request was created for this session. Call after successfully creating a PR."
}
func (t *reportPRCreatedTool) InputSchema() json.RawMessage {
	return json.RawMessage(`{
		"type": "object",
		"properties": {
			"prNumber": {"type": "number", "description": "The PR number"},
			"prUrl": {"type": "string", "description": "The full PR URL (https://github.com/owner/repo/pull/NNN)"}
		},
		"required": ["prNumber", "prUrl"]
	}`)
}

func (t *reportPRCreatedTool) Execute(ctx context.Context, input json.RawMessage) (*tool.Result, error) {
	var params struct {
		PRNumber int    `json:"prNumber"`
		PRURL    string `json:"prUrl"`
	}
	if err := json.Unmarshal(input, &params); err != nil {
		return tool.ErrorResult(fmt.Sprintf("invalid input: %v", err)), nil
	}
	if params.PRNumber <= 0 || params.PRURL == "" {
		return tool.ErrorResult("prNumber and prUrl are required"), nil
	}
	if !prURLPattern.MatchString(params.PRURL) {
		return tool.ErrorResult("prUrl must match https://github.com/owner/repo/pull/NNN"), nil
	}

	if t.svc.PRWatcher == nil {
		return tool.ErrorResult("PR tracking not yet initialized — please retry in a moment"), nil
	}
	t.svc.PRWatcher.RegisterPRFromAgent(t.ctx.SessionID, params.PRNumber, params.PRURL)

	return &tool.Result{Content: fmt.Sprintf("PR #%d reported and tracked: %s", params.PRNumber, params.PRURL)}, nil
}

// --- report_pr_merged ---

type reportPRMergedTool struct {
	svc *Services
	ctx *ToolContext
}

func (t *reportPRMergedTool) Name() string           { return "mcp__chatml__report_pr_merged" }
func (t *reportPRMergedTool) IsConcurrentSafe() bool { return true }
func (t *reportPRMergedTool) DeferLoading() bool     { return true }
func (t *reportPRMergedTool) Description() string {
	return "Report that a pull request was merged for this session."
}
func (t *reportPRMergedTool) InputSchema() json.RawMessage {
	return json.RawMessage(`{
		"type": "object",
		"properties": {
			"prNumber": {"type": "number", "description": "The PR number that was merged (optional)"}
		}
	}`)
}

func (t *reportPRMergedTool) Execute(ctx context.Context, input json.RawMessage) (*tool.Result, error) {
	if t.svc.PRWatcher == nil {
		return tool.ErrorResult("PR tracking not yet initialized — please retry in a moment"), nil
	}
	t.svc.PRWatcher.ForceCheckSession(t.ctx.SessionID)
	return &tool.Result{Content: "PR merge reported. Session status will be updated."}, nil
}

// --- clear_pr_link ---

type clearPRLinkTool struct {
	svc *Services
	ctx *ToolContext
}

func (t *clearPRLinkTool) Name() string           { return "mcp__chatml__clear_pr_link" }
func (t *clearPRLinkTool) IsConcurrentSafe() bool { return true }
func (t *clearPRLinkTool) DeferLoading() bool     { return true }
func (t *clearPRLinkTool) Description() string {
	return "Clear the pull request link from this session."
}
func (t *clearPRLinkTool) InputSchema() json.RawMessage {
	return json.RawMessage(`{"type": "object", "properties": {}}`)
}

func (t *clearPRLinkTool) Execute(ctx context.Context, input json.RawMessage) (*tool.Result, error) {
	if t.svc.PRWatcher == nil {
		return tool.ErrorResult("PR tracking not yet initialized — please retry in a moment"), nil
	}
	t.svc.PRWatcher.UnlinkPR(t.ctx.SessionID)
	return &tool.Result{Content: "PR link cleared from session."}, nil
}
